pkg/git/libgit2/checkout.go

```diff
@@ -161,19 +161,44 @@ func (c *CheckoutBranch) Checkout(ctx context.Context, path, url string, opts *g
 		}
 		defer upstreamCommit.Free()
 
-		// Once the index has been updated with Fetch, and we know the tip commit,
-		// a hard reset can be used to align the local worktree with the remote branch's.
-		err = repo.ResetToCommit(upstreamCommit, git2go.ResetHard, &git2go.CheckoutOptions{
+		localBranch, err := repo.LookupBranch(c.Branch, git2go.BranchLocal)
+		if err != nil {
+			// We ignore the error if the branch is not found since we are going to
+			// create it ourselves.
+			if !git2go.IsErrorCode(err, git2go.ErrorCodeNotFound) {
+				return nil, fmt.Errorf("cannot lookup branch '%s': %w", c.Branch, err)
+			}
+			localBranch, err = repo.CreateBranch(c.Branch, upstreamCommit, false)
+			if err != nil {
+				return nil, fmt.Errorf("cannot create local branch '%s': %w", c.Branch, err)
+			}
+		}
+		defer localBranch.Free()
+
+		tree, err := repo.LookupTree(upstreamCommit.TreeId())
+		if err != nil {
+			return nil, fmt.Errorf("cannot lookup tree for branch '%s': %w", c.Branch, err)
```

Comment:
All the error messages introduced in this change can start with "unable to..." similar to the rest of code surrounding it.